File: 4-AnagramGroups/solution.java
import java.util.*;

class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        // Mapa para agrupar os anagramas
        Map<String, List<String>> map = new HashMap<>();
        
        // Itera por todas as palavras
        for (String str : strs) {
            // Converte a string em um array de caracteres, ordena e volta a ser string
            char[] charArray = str.toCharArray();
            Arrays.sort(charArray);
            String sorted = new String(charArray);
            
            // Se já existir a chave, adiciona a palavra no grupo
            if (!map.containsKey(sorted)) {
                map.put(sorted, new ArrayList<>());
            }
            map.get(sorted).add(str);
        }
        
        // Retorna os valores do mapa como uma lista de listas
        return new ArrayList<>(map.values());
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        String[] strs = {"act", "pots", "tops", "cat", "stop", "hat"};
        
        // Testa a função e imprime o resultado
        List<List<String>> result = solution.groupAnagrams(strs);
        System.out.println(result);
    }
}

File: 4-AnagramGroups/solution.js
class Solution {
    /**
     * @param {string[]} strs
     * @return {string[][]}
     */
    groupAnagrams(strs) {
        // Cria um mapa para agrupar os anagramas
        const map = new Map()

        for (const str of strs) {
             // Ordena as letras da palavra
            const sorted = str.split('').sort().join('');
            if (map.has(sorted)) {
                map.get(sorted).push(str);
            } else {
                // Se não, cria uma nova lista para esse anagrama
                map.set(sorted, [str]);
            }
        }
        // Retorna os grupos de anagramas como um array de arrays
        return Array.from(map.values());
    }
}

const solution = new Solution();
const strs = ["act", "pots", "tops", "cat", "stop", "hat"];
console.log(solution.groupAnagrams(strs));